File: app/src/main/java/ru/voronezhtsev/weatherapp/data/db/ForecastsDAO.java
package ru.voronezhtsev.weatherapp.data.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.List;

import ru.voronezhtsev.weatherapp.models.data.db.Forecast;

public class ForecastsDAO extends SQLiteOpenHelper {

    private static final String TABLE_NAME = "forecasts";
    private static final String DATETIME_COLUMN = "datetime";
    private static final String TEMP_COLUMN = "temperature";
    private static final String TEMP_MIN_COLUMN = "temperature_min";
    private static final String TEMP_MAX_COLUMN = "temperature_max";
    private static final String TAG = "ForecastsDAO";
    private static final String ERR_ADD_FORECAST = "Error while adding forecast with datetime: ";
    private static final int VERSION = 1;
    private static final String NAME = "forecasts.db";

    private static ForecastsDAO INSTANCE = null;
    //todo Растет БД с каждым запуском приложения
    private static final String CREATE_TABLE =
            "create table if not exists " +
                    TABLE_NAME + "("
                    + DATETIME_COLUMN  + " INTEGER, "
                    + TEMP_COLUMN +" REAL, "
                    + TEMP_MIN_COLUMN + " REAL, "
                    + TEMP_MAX_COLUMN+ " REAL)";
    private ForecastsDAO(Context context) {
        super(context, NAME, null, VERSION);
    }

    public static synchronized ForecastsDAO getInstance(Context context) {
        if(INSTANCE == null) {
            INSTANCE = new ForecastsDAO(context);
        }
        return INSTANCE;
    }
    public void addForecasts(List<Forecast> forecasts) {
        SQLiteDatabase database = getWritableDatabase();
        for(Forecast forecast: forecasts) {
            ContentValues cv = getContentValues(forecast);
            database.insertOrThrow(TABLE_NAME, null, cv);
        }
        database.close();
    }

    public List<Forecast> getForecasts() {
        SQLiteDatabase database = getReadableDatabase();
        Cursor cursor = database.query(TABLE_NAME,
                new String[] {
                        DATETIME_COLUMN,
                        TEMP_COLUMN,
                        TEMP_MAX_COLUMN,
                        TEMP_MIN_COLUMN
                }, null, null, null, null, null);

        List<Forecast> forecasts = new ArrayList<>();
        if(cursor.moveToFirst()) {
            int dateTimeColIdx = cursor.getColumnIndex(DATETIME_COLUMN);
            int tempColIdx = cursor.getColumnIndex(TEMP_COLUMN);
            int temMinColIdx = cursor.getColumnIndex(TEMP_MIN_COLUMN);
            int tempMaxColIdx = cursor.getColumnIndex(TEMP_MAX_COLUMN);

            while (!cursor.isAfterLast()) {
                long dateTime = cursor.getLong(dateTimeColIdx);
                double temp = cursor.getDouble(tempColIdx);
                double tempMin = cursor.getDouble(temMinColIdx);
                double tempMax = cursor.getDouble(tempMaxColIdx);
                Forecast forecast = new Forecast(dateTime, temp, tempMin, tempMax);
                forecasts.add(forecast);
                cursor.moveToNext();
            }
        }
        return forecasts;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(CREATE_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

    }
    private ContentValues getContentValues(Forecast forecast) {
        ContentValues cv = new ContentValues();
        cv.put(DATETIME_COLUMN, forecast.getDateTime());
        cv.put(TEMP_COLUMN, forecast.getTemp());
        cv.put(TEMP_MIN_COLUMN, forecast.getTempMin());
        cv.put(TEMP_MAX_COLUMN, forecast.getTempMax());
        return cv;
    }
}

File: app/src/main/java/ru/voronezhtsev/weatherapp/data/db/ForecastsDAO.kt
package ru.voronezhtsev.weatherapp.data.db

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import ru.voronezhtsev.weatherapp.di.SingletonHolder
import ru.voronezhtsev.weatherapp.models.data.network.ForecastResponse
import ru.voronezhtsev.weatherapp.models.domain.Forecast

class ForecastsDAO private constructor(context: Context) : SQLiteOpenHelper(context, "forecasts.db", null, 1) {
    //todo На экране выводится прогноз на текущие 24 часа, а хранится на 5 дней для каждого города
    private val DATABASE_DDL = "create table if not exists FORECASTS( " +
            "CITY_NAME   text, " +
            "TEMPERATURE real, " +
            "ICON_CODE   text, " +
            "DATETIME    integer, " +
            "DESCRIPTION text)"

    companion object : SingletonHolder<ForecastsDAO, Context>(::ForecastsDAO)

    override fun onCreate(db: SQLiteDatabase?) {
        db?.execSQL(DATABASE_DDL)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
    }

    fun save(response: ForecastResponse) {
        for (contentValues in getContentValues(response)) {
            writableDatabase.replaceOrThrow("FORECASTS", null, contentValues)
        }
    }

    private fun getContentValues(response: ForecastResponse): List<ContentValues> {
        val result = mutableListOf<ContentValues>()
        for (forecast in response.forecast) {
            val contentValues = ContentValues()
            contentValues.put("CITY_NAME", response.city.name)
            contentValues.put("TEMPERATURE", forecast.main.temp - 273.15)
            contentValues.put("ICON_CODE", forecast.weather[0].icon)
            contentValues.put("DATETIME", forecast.dateTime)
            contentValues.put("DESCRIPTION", forecast.weather[0].description)
            result.add(contentValues)
        }
        return result
    }

    fun load(cityName: String): List<Forecast> {
        val cursor = readableDatabase.query("FORECASTS", null, "CITY_NAME = ?",
                arrayOf(cityName), null, null, "DATETIME")
        val result = mutableListOf<Forecast>()
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast) {
                val cityName = cursor.getString(cursor.getColumnIndex("CITY_NAME"));
                val temperature = cursor.getDouble(cursor.getColumnIndex("TEMPERATURE"));
                val iconCode = cursor.getString(cursor.getColumnIndex("ICON_CODE"));
                val dateTime = cursor.getLong(cursor.getColumnIndex("DATETIME"))
                val description = cursor.getString(cursor.getColumnIndex("DESCRIPTION"))
                val forecast = Forecast(cityName, dateTime, temperature, iconCode, description)
                result.add(forecast)
                cursor.moveToNext()
            }
        }
        cursor.close()
        return result
    }
}